### PR TITLE
feat: turbofish in struct pattern

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Part of https://github.com/noir-lang/noir/issues/5584

## Summary

Allows parsing turbofish in a struct pattern path segments, then uses those turbofish generics for type binding/inference.

## Additional Context

None.

## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/tests.rs [hunk 1]
```diff
@@ -2610,3 +2610,58 @@ fn turbofish_in_middle_of_variable_unsupported_yet() {
         CompilationError::TypeError(TypeCheckError::UnsupportedTurbofishUsage { .. }),
     ));
 }
+
+#[test]
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/tests.rs:2614]: Please add a test with zero expected errors

File: compiler/noirc_frontend/src/parser/parser/primitives.rs [hunk 2]
```diff
@@ -96,7 +100,7 @@ pub(super) fn turbofish<'a>(
 }
 
 pub(super) fn variable() -> impl NoirParser<ExpressionKind> {
-    path().map(ExpressionKind::Variable)
+    path(super::parse_type()).map(ExpressionKind::Variable)
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/parser/parser/primitives.rs:103]: Can any variable now have turbofish, e.g. `let x: u8 = 3; assert(x::<> == 3)`?

File: compiler/noirc_frontend/src/elaborator/patterns.rs [hunk 3]
```diff
@@ -178,6 +182,13 @@ impl<'context> Elaborator<'context> {
             }
         };
 
+        let generics = self.resolve_struct_turbofish_generics(
+            &struct_type.borrow(),
+            generics,
+            last_segment.generics,
+            Span::from(last_segment.ident.span().end()..last_segment.span.end()),
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/elaborator/patterns.rs:189]: Empty span:
```suggestion
            Span::from(last_segment.ident.span().start()..last_segment.span.end()),
```

**Review by michaeljklein — CHANGES_REQUESTED** [review]

File: compiler/noirc_frontend/src/parser/parser/primitives.rs (hunk 2 above)
Comment (asterite) [reply compiler/noirc_frontend/src/parser/parser/primitives.rs:103]: Oh, this was also the case before this PR. There's also `variable_no_turbofish`. Or at least turbofish at the end of a variable was allowed before this PR and before we allowed it in the middle of a path (though I think it's ignored if it's not used as a function).

File: compiler/noirc_frontend/src/elaborator/patterns.rs (hunk 3 above)
Comment (asterite) [reply compiler/noirc_frontend/src/elaborator/patterns.rs:189]: Ah, sorry, this is confusing 😅 

The idea is to have the error point like this:

```
foo::<T>
   ^~~~~
```

Since I used a similar code in another place I'll define a function on `PathSegment` for this, and also document it.

File: compiler/noirc_frontend/src/parser/parser/primitives.rs (hunk 2 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/parser/parser/primitives.rs:103]: Rust allows the above btw

File: compiler/noirc_frontend/src/elaborator/patterns.rs (hunk 3 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/elaborator/patterns.rs:189]: The span isn't empty, the intent is to underline only the turbofish and not the variable as well

File: compiler/noirc_frontend/src/parser/parser/types.rs [hunk 4]
```diff
@@ -181,9 +181,9 @@ pub(super) fn int_type() -> impl NoirParser<UnresolvedType> {
 pub(super) fn named_type<'a>(
     type_parser: impl NoirParser<UnresolvedType> + 'a,
 ) -> impl NoirParser<UnresolvedType> + 'a {
-    path_no_turbofish().then(generic_type_args(type_parser)).map_with_span(|(path, args), span| {
-        UnresolvedTypeData::Named(path, args, false).with_span(span)
-    })
+    path(type_parser.clone()).then(generic_type_args(type_parser)).map_with_span(
+        |(path, args), span| UnresolvedTypeData::Named(path, args, false).with_span(span),
+    )
```
Comment (jfecher) [inline compiler/noirc_frontend/src/parser/parser/types.rs:186]: This would allow `Foo::<Bar><Baz>`, right? We should probably error in that case

**Review by jfecher — CHANGES_REQUESTED** [review]
Found a bit of syntax we shouldn't allow

File: compiler/noirc_frontend/src/parser/parser/types.rs (hunk 4 above)
Comment (asterite) [reply compiler/noirc_frontend/src/parser/parser/types.rs]: Good point! It turns out `path_no_turbofish()` here was fine. For some reason I thought this is where struct patterns are parsed, but that happens in a different place.

**Review by michaeljklein — APPROVED** [review]
LGTM
